File: qqbot-module-platform-admin/src/main/java/com/mieai/qqbot/admin/security/AdminLoginAttemptGuard.java
package com.mieai.qqbot.admin.security;

import com.mieai.qqbot.persistence.internal.DatabaseExceptionClassifier;
import com.mieai.qqbot.persistence.internal.UtcTimestampCodec;
import java.nio.charset.StandardCharsets;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.time.Clock;
import java.time.Duration;
import java.time.Instant;
import java.time.temporal.ChronoUnit;
import java.util.HashMap;
import java.util.HexFormat;
import java.util.Map;
import javax.sql.DataSource;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.dao.DataAccessException;
import org.springframework.jdbc.core.JdbcTemplate;
import org.springframework.jdbc.datasource.DataSourceTransactionManager;
import org.springframework.stereotype.Component;
import org.springframework.transaction.support.TransactionTemplate;

/** Shared JDBC login throttle; tests may use the in-memory fallback constructor. */
@Component
final class AdminLoginAttemptGuard {
    private static final int MAX_FAILURES = 5;
    private static final int MAX_TRACKED_KEYS = 10_000;
    private static final Duration BLOCK_DURATION = Duration.ofMinutes(15);
    private static final Duration RETENTION = Duration.ofHours(1);

    private final Map<String, Attempt> attempts = new HashMap<>();
    private final Clock clock;
    private final JdbcTemplate jdbc;
    private final TransactionTemplate transaction;

    @Autowired
    AdminLoginAttemptGuard(DataSource dataSource) {
        this(dataSource, Clock.systemUTC());
    }

    AdminLoginAttemptGuard(DataSource dataSource, Clock clock) {
        this.clock = clock;
        this.jdbc = new JdbcTemplate(dataSource);
        this.transaction = new TransactionTemplate(new DataSourceTransactionManager(dataSource));
    }

    AdminLoginAttemptGuard() {
        this(Clock.systemUTC());
    }

    AdminLoginAttemptGuard(Clock clock) {
        this.clock = clock;
        this.jdbc = null;
        this.transaction = null;
    }

    synchronized void check(String remoteAddress, String username) {
        if (jdbc != null) { checkJdbc(remoteAddress, username); return; }
        String key = key(remoteAddress, username);
        Attempt attempt = attempts.get(key);
        if (attempt == null) {
            return;
        }
        Instant now = clock.instant();
        if (attempt.blockedUntil() != null && attempt.blockedUntil().isAfter(now)) {
            throw throttled(now, attempt.blockedUntil());
        }
        if (attempt.blockedUntil() != null) {
            attempts.remove(key);
        }
    }

    synchronized void failed(String remoteAddress, String username) {
        if (jdbc != null) { failedJdbc(remoteAddress, username); return; }
        String key = key(remoteAddress, username);
        Instant now = clock.instant();
        Attempt previous = attempts.get(key);
        int failures = previous == null ? 1 : previous.failures() + 1;
        Instant blockedUntil = failures >= MAX_FAILURES ? now.plus(BLOCK_DURATION) : null;
        attempts.put(key, new Attempt(failures, blockedUntil, now));
        prune(now);
        if (blockedUntil != null) {
            throw throttled(now, blockedUntil);
        }
    }

    synchronized void succeeded(String remoteAddress, String username) {
        if (jdbc != null) {
            jdbc.update("DELETE FROM admin_login_attempts WHERE attempt_key=?", keyHash(remoteAddress, username));
            return;
        }
        attempts.remove(key(remoteAddress, username));
    }

    private void checkJdbc(String remoteAddress, String username) {
        String key = keyHash(remoteAddress, username);
        AttemptRow row = jdbc.query("SELECT failures, blocked_until FROM admin_login_attempts WHERE attempt_key=?",
                (rs, n) -> new AttemptRow(rs.getInt(1), rs.getString(2)), key)
                .stream().findFirst().orElse(null);
        if (row == null || row.blockedUntil() == null) return;
        Instant blocked = UtcTimestampCodec.parse(row.blockedUntil());
        Instant now = clock.instant();
        if (blocked.isAfter(now)) throw throttled(now, blocked);
        jdbc.update("DELETE FROM admin_login_attempts WHERE attempt_key=?", key);
    }

    private void failedJdbc(String remoteAddress, String username) {
        String key = keyHash(remoteAddress, username);
        Instant now = clock.instant();
        Instant blocked = now.plus(BLOCK_DURATION);
        transaction.executeWithoutResult(status -> {
            jdbc.update("DELETE FROM admin_login_attempts WHERE last_touched < ?",
                    UtcTimestampCodec.format(now.minus(RETENTION)));
            int updated = jdbc.update("""
                    UPDATE admin_login_attempts
                    SET failures=failures+1,
                        blocked_until=CASE WHEN failures+1 >= ? THEN ? ELSE blocked_until END,
                        last_touched=?
                    WHERE attempt_key=?
                    """, MAX_FAILURES, UtcTimestampCodec.format(blocked), UtcTimestampCodec.format(now), key);
            if (updated == 0) {
                try {
                    jdbc.update("INSERT INTO admin_login_attempts(attempt_key, failures, blocked_until, last_touched) VALUES (?, 1, NULL, ?)",
                            key, UtcTimestampCodec.format(now));
                } catch (DataAccessException race) {
                    if (!DatabaseExceptionClassifier.isDuplicateKey(race)) throw race;
                    jdbc.update("""
                            UPDATE admin_login_attempts
                            SET failures=failures+1,
                                blocked_until=CASE WHEN failures+1 >= ? THEN ? ELSE blocked_until END,
                                last_touched=?
                            WHERE attempt_key=?
                            """, MAX_FAILURES, UtcTimestampCodec.format(blocked),
                            UtcTimestampCodec.format(now), key);
                }
            }
        });
        AttemptRow row = jdbc.query("SELECT failures, blocked_until FROM admin_login_attempts WHERE attempt_key=?",
                (rs, n) -> new AttemptRow(rs.getInt(1), rs.getString(2)), key)
                .stream().findFirst().orElse(null);
        if (row != null && row.failures() >= MAX_FAILURES) {
            Instant until = row.blockedUntil() == null ? blocked : UtcTimestampCodec.parse(row.blockedUntil());
            throw throttled(now, until);
        }
    }

    private void prune(Instant now) {
        if (attempts.size() <= MAX_TRACKED_KEYS) {
            return;
        }
        Instant cutoff = now.minus(RETENTION);
        attempts.entrySet().removeIf(entry -> entry.getValue().lastTouched().isBefore(cutoff));
        while (attempts.size() > MAX_TRACKED_KEYS) {
            String oldest = attempts.entrySet().stream()
                    .min(Map.Entry.comparingByValue(
                            java.util.Comparator.comparing(Attempt::lastTouched)))
                    .map(Map.Entry::getKey)
                    .orElse(null);
            if (oldest == null) {
                return;
            }
            attempts.remove(oldest);
        }
    }

    private static LoginThrottledException throttled(Instant now, Instant blockedUntil) {
        long seconds = now.until(blockedUntil, ChronoUnit.SECONDS);
        return new LoginThrottledException(seconds);
    }

    private static String key(String remoteAddress, String username) {
        String address = remoteAddress == null || remoteAddress.isBlank() ? "unknown" : remoteAddress;
        return address + '\0' + username;
    }

    private static String keyHash(String remoteAddress, String username) {
        String value = key(remoteAddress, username);
        try {
            return HexFormat.of().formatHex(MessageDigest.getInstance("SHA-256")
                    .digest(value.getBytes(StandardCharsets.UTF_8)));
        } catch (NoSuchAlgorithmException exception) {
            throw new IllegalStateException("SHA-256 is unavailable", exception);
        }
    }

    private record Attempt(int failures, Instant blockedUntil, Instant lastTouched) {}
    private record AttemptRow(int failures, String blockedUntil) {}
}

File: qqbot-module-platform-admin/src/main/kotlin/com/mieai/qqbot/admin/security/AdminLoginAttemptGuard.kt
package com.mieai.qqbot.admin.security

import com.mieai.qqbot.persistence.internal.DatabaseExceptionClassifier
import com.mieai.qqbot.persistence.internal.UtcTimestampCodec
import java.nio.charset.StandardCharsets
import java.security.MessageDigest
import java.security.NoSuchAlgorithmException
import java.time.Clock
import java.time.Duration
import java.time.Instant
import java.time.temporal.ChronoUnit
import java.util.HexFormat
import javax.sql.DataSource
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.dao.DataAccessException
import org.springframework.jdbc.core.JdbcTemplate
import org.springframework.jdbc.core.RowMapper
import org.springframework.jdbc.datasource.DataSourceTransactionManager
import org.springframework.stereotype.Component
import org.springframework.transaction.support.TransactionTemplate

/** Shared JDBC login throttle; tests may use the in-memory fallback constructor. */
@Component
class AdminLoginAttemptGuard private constructor(
    private val clock: Clock,
    private val jdbc: JdbcTemplate?,
    private val transaction: TransactionTemplate?,
) {
    private val attempts = HashMap<String, Attempt>()

    @Autowired
    constructor(dataSource: DataSource) : this(dataSource, Clock.systemUTC())

    constructor(dataSource: DataSource, clock: Clock) : this(
        clock,
        JdbcTemplate(dataSource),
        TransactionTemplate(DataSourceTransactionManager(dataSource)),
    )

    constructor(clock: Clock) : this(clock, null, null)

    @Synchronized
    fun check(remoteAddress: String?, username: String) {
        if (jdbc != null) {
            checkJdbc(remoteAddress, username)
            return
        }
        val key = key(remoteAddress, username)
        val attempt = attempts[key] ?: return
        val now = clock.instant()
        val blockedUntil = attempt.blockedUntil
        if (blockedUntil != null && blockedUntil.isAfter(now)) throw throttled(now, blockedUntil)
        if (blockedUntil != null) attempts.remove(key)
    }

    @Synchronized
    fun failed(remoteAddress: String?, username: String) {
        if (jdbc != null) {
            failedJdbc(remoteAddress, username)
            return
        }
        val key = key(remoteAddress, username)
        val now = clock.instant()
        val failures = (attempts[key]?.failures ?: 0) + 1
        val blockedUntil = if (failures >= MAX_FAILURES) now.plus(BLOCK_DURATION) else null
        attempts[key] = Attempt(failures, blockedUntil, now)
        prune(now)
        if (blockedUntil != null) throw throttled(now, blockedUntil)
    }

    @Synchronized
    fun succeeded(remoteAddress: String?, username: String) {
        val template = jdbc
        if (template != null) {
            template.update(
                "DELETE FROM admin_login_attempts WHERE attempt_key=?",
                keyHash(remoteAddress, username),
            )
            return
        }
        attempts.remove(key(remoteAddress, username))
    }

    private fun checkJdbc(remoteAddress: String?, username: String) {
        val template = requireNotNull(jdbc)
        val key = keyHash(remoteAddress, username)
        val row = template.query(
            "SELECT failures, blocked_until FROM admin_login_attempts WHERE attempt_key=?",
            ATTEMPT_ROW_MAPPER,
            key,
        ).firstOrNull()
        val blockedText = row?.blockedUntil ?: return
        val blocked = UtcTimestampCodec.parse(blockedText)
        val now = clock.instant()
        if (blocked.isAfter(now)) throw throttled(now, blocked)
        template.update("DELETE FROM admin_login_attempts WHERE attempt_key=?", key)
    }

    private fun failedJdbc(remoteAddress: String?, username: String) {
        val template = requireNotNull(jdbc)
        val transactionTemplate = requireNotNull(transaction)
        val key = keyHash(remoteAddress, username)
        val now = clock.instant()
        val blocked = now.plus(BLOCK_DURATION)
        transactionTemplate.executeWithoutResult {
            template.update(
                "DELETE FROM admin_login_attempts WHERE last_touched < ?",
                UtcTimestampCodec.format(now.minus(RETENTION)),
            )
            val updated = template.update(
                """
                UPDATE admin_login_attempts
                SET failures=failures+1,
                    blocked_until=CASE WHEN failures+1 >= ? THEN ? ELSE blocked_until END,
                    last_touched=?
                WHERE attempt_key=?
                """.trimIndent(),
                MAX_FAILURES,
                UtcTimestampCodec.format(blocked),
                UtcTimestampCodec.format(now),
                key,
            )
            if (updated == 0) {
                try {
                    template.update(
                        "INSERT INTO admin_login_attempts(attempt_key, failures, blocked_until, last_touched) " +
                            "VALUES (?, 1, NULL, ?)",
                        key,
                        UtcTimestampCodec.format(now),
                    )
                } catch (race: DataAccessException) {
                    if (!DatabaseExceptionClassifier.isDuplicateKey(race)) throw race
                    template.update(
                        """
                        UPDATE admin_login_attempts
                        SET failures=failures+1,
                            blocked_until=CASE WHEN failures+1 >= ? THEN ? ELSE blocked_until END,
                            last_touched=?
                        WHERE attempt_key=?
                        """.trimIndent(),
                        MAX_FAILURES,
                        UtcTimestampCodec.format(blocked),
                        UtcTimestampCodec.format(now),
                        key,
                    )
                }
            }
        }
        val row = template.query(
            "SELECT failures, blocked_until FROM admin_login_attempts WHERE attempt_key=?",
            ATTEMPT_ROW_MAPPER,
            key,
        ).firstOrNull()
        if (row != null && row.failures >= MAX_FAILURES) {
            val until = row.blockedUntil?.let(UtcTimestampCodec::parse) ?: blocked
            throw throttled(now, until)
        }
    }

    private fun prune(now: Instant) {
        if (attempts.size <= MAX_TRACKED_KEYS) return
        val cutoff = now.minus(RETENTION)
        attempts.entries.removeAll { it.value.lastTouched.isBefore(cutoff) }
        while (attempts.size > MAX_TRACKED_KEYS) {
            val oldest = attempts.entries.minByOrNull { it.value.lastTouched }?.key ?: return
            attempts.remove(oldest)
        }
    }

    private data class Attempt(
        val failures: Int,
        val blockedUntil: Instant?,
        val lastTouched: Instant,
    )

    private data class AttemptRow(
        val failures: Int,
        val blockedUntil: String?,
    )

    private companion object {
        const val MAX_FAILURES = 5
        const val MAX_TRACKED_KEYS = 10_000
        val BLOCK_DURATION: Duration = Duration.ofMinutes(15)
        val RETENTION: Duration = Duration.ofHours(1)
        val ATTEMPT_ROW_MAPPER: RowMapper<AttemptRow> = RowMapper { resultSet, _ ->
            AttemptRow(resultSet.getInt(1), resultSet.getString(2))
        }

        fun throttled(now: Instant, blockedUntil: Instant): LoginThrottledException =
            LoginThrottledException(now.until(blockedUntil, ChronoUnit.SECONDS))

        fun key(remoteAddress: String?, username: String): String {
            val address = remoteAddress?.takeIf { it.isNotBlank() } ?: "unknown"
            return "$address\u0000$username"
        }

        fun keyHash(remoteAddress: String?, username: String): String {
            val value = key(remoteAddress, username)
            val digest = try {
                MessageDigest.getInstance("SHA-256")
            } catch (error: NoSuchAlgorithmException) {
                throw IllegalStateException("SHA-256 is unavailable", error)
            }
            return HexFormat.of().formatHex(digest.digest(value.toByteArray(StandardCharsets.UTF_8)))
        }
    }
}
